packages/devtools_app/lib/src/debugger/flutter/codeview.dart

```diff
@@ -22,57 +22,92 @@ import 'debugger_model.dart';
 class CodeView extends StatefulWidget {
   const CodeView({
     Key key,
-    this.script,
-    this.stack,
     this.controller,
+    this.scriptRef,
     this.onSelected,
-    this.lineNumberToBreakpoint,
   }) : super(key: key);
 
   static const rowHeight = 20.0;
   static const assumedCharacterWidth = 16.0;
 
   final DebuggerController controller;
-  final Map<int, BreakpointAndSourcePosition> lineNumberToBreakpoint;
-  final Script script;
-  final vm.Stack stack;
-  final void Function(Script script, int line) onSelected;
+  final ScriptRef scriptRef;
+
+  final void Function(ScriptRef scriptRef, int line) onSelected;
 
   @override
   _CodeViewState createState() => _CodeViewState();
 }
 
 class _CodeViewState extends State<CodeView> {
+  Script script;
   List<String> lines = [];
   Set<int> executableLines = {};
+
   LinkedScrollControllerGroup verticalController;
   ScrollController gutterController;
   ScrollController textController;
 
-  // The paused positions in the current [widget.script] from the [widget.stack].
-  List<int> pausedPositions;
+  ScriptRef get scriptRef => widget.scriptRef;
 
   @override
   void initState() {
     super.initState();
 
+    _initScriptInfo();
+
     verticalController = LinkedScrollControllerGroup();
     gutterController = verticalController.addAndGet();
     textController = verticalController.addAndGet();
-    _updateLines();
-    _updatePausedPositions();
+
+    widget.controller.scriptLocation.addListener(_handleScriptLocationChanged);
+  }
+
+  void _initScriptInfo() {
+    script = widget.controller.getScriptCached(scriptRef);
+
+    if (script == null) {
+      if (scriptRef != null) {
+        widget.controller.getScript(scriptRef).then((script) {
+          setState(() {
+            this.script = script;
+
+            _parseScriptLines();
+          });
+        });
+      }
+    } else {
+      _parseScriptLines();
+    }
+  }
+
+  void _parseScriptLines() {
+    lines = script.source?.split('\n') ?? [];
+
+    // TODO(devoncarew): Change to using SourceReportRange.possibleBreakpoints.
+    // (see getSourceReport('PossibleBreakpoints').
+    executableLines = {};
+    // Recalculate the executable lines.
+    if (script.tokenPosTable != null) {
+      for (var encodedInfo in script.tokenPosTable) {
+        executableLines.add(encodedInfo[0]);
+      }
+    }
   }
 
   @override
   void didUpdateWidget(CodeView oldWidget) {
     super.didUpdateWidget(oldWidget);
 
-    if (widget.script != oldWidget.script) {
-      _updateLines();
+    if (widget.controller != oldWidget.controller) {
+      oldWidget.controller.scriptLocation
```

Comment:
use autoDispose here to avoid having to manually remove the old listener assuming that all old listeners can be disposed before adding new ones.

Reply:
Added a todo: to investigate this. There are some other things we're disposing, and I still need to get a better handle on the widget lifecycle and the AutoDispose mixin.

Reply:
If use addAutoDisposeListener then the listener is automatically disposed when dispose on the widget is called. You can additionally dispose all existing listeners typically in `didChangeDependencies` by calling `cancel` manually.
There is a good example from the logging view.
https://github.com/flutter/devtools/blob/master/packages/devtools_app/lib/src/logging/flutter/logging_screen.dart#L88

https://github.com/flutter/devtools/blob/master/packages/devtools_app/lib/src/logging/flutter/logging_screen.dart#L88

Reply:
OK, I re-read the code and that example, and updated this to use AutoDisposeMixin.